source/git/source.go

```diff
@@ -349,10 +354,18 @@ func (gs *gitSourceHandler) CacheKey(ctx context.Context, g session.Group, index
 	gs.locker.Lock(remote)
 	defer gs.locker.Unlock(remote)
 
-	if ref := gs.src.Ref; ref != "" && gitutil.IsCommitSHA(ref) {
-		cacheKey := gs.shaToCacheKey(ref, "")
+	var refCommitFullHash string
+	if gitutil.IsCommitSHA(gs.src.CommitHash) {
+		refCommitFullHash = gs.src.CommitHash
+	}
+	if refCommitFullHash == "" && gitutil.IsCommitSHA(gs.src.Ref) {
+		refCommitFullHash = gs.src.Ref
+	}
+	if refCommitFullHash != "" {
+		cacheKey := gs.shaToCacheKey(refCommitFullHash, "")
```

Comment:
The ref should be set here if one is present even if we know the full hash.

If first build is from `ref=master,commithash=` and second build(replay pinned from provenance) is from `ref=master,commithash=abcdef` then they should generate same cache keys. Ideally we would cover this with a testcase.